### ISSUE TITLE
The link would be opened when drag and drop it in the Tab view on Ubuntu 18.04

### ISSUE THREAD

**Issue by qqlovekerry** [josh-berry/tab-stash#51]
### Describe the Bug
*A clear and concise description of what the problem is.*
When the links in the rearranged in the Tab view, the link will be opened in the browser. I think this is unwanted and such behavior does not show if the link is dragged and dropped in the sidebar view.

### Steps to Reproduce
*Detailed steps to reproduce the problem:*
1. Open Tab Stash with the Tab view.
2. Drag and drop a link to move it to another group.
3. The link will be rearranged but it is also opened in the browser.

### Expected Behavior
*A clear and concise description of what you expected to happen.*
If the saved link is rearranged in the sidebar view, the function works great. And the expected behavior is that the link is moved without opened in Firefox if the Tab view if used.

### Screenshots
*If applicable, add screenshots to help explain your problem.*

### Desktop (please complete the following information):
 - **OS:** Ububutu 18.04
 - **Firefox Browser Version:** *75.0*

**Comment by josh-berry** [comment]
Hi, I'm afraid I haven't been able to reproduce this behavior at all in Firefox 75 on Linux (I have Ubuntu 19.10 on my side). It seems especially strange that you're only hitting this issue with the tab view (and not the sidebar view), since they both use the same code underneath with different styling applied. Can you share a little more info about your setup?

* Which desktop environment are you using? Just the default Ubuntu environment or do you have a variant like KDE or XFCE? (And do you know if you are using X or Wayland?)
* Is this a VM or a physical machine? If it's a VM, what VM software are you using (VirtualBox? VMware? Hyper-V? Something else?)

The reason I ask about the VM is I have noticed some VM engines use different methods for sending mouse input to the VM, and this can sometimes cause oddities with drag and drop and mouse-over events. That said, I'm still mystified as to why you would only be seeing this problem in the tab view and not the sidebar...

**Comment by qqlovekerry** [comment]
Hi, thanks for your reply! You do not face the same problem on Ubuntu 19.10? Is this only related to 18.04?
I am just an end-user of Ubuntu and I will try my best to provide as much information as possible.
- I am using Gnome desktop environment on Ubuntu, I think this is the default of Ubuntu 18.04. I think I am using Wayland but I am not quite sure. Could you please tell how to check that?
- I am using a physical machine not a virtual machine.

**Comment by josh-berry** [comment]
If you are using the default environment of Ubuntu, that tells me enough to try to replicate your setup, so I will try to do that and see if I can reproduce the problem. (That said, I will only be able to do it as a virtual machine, but hopefully that's close enough.) I can also find out whether it's Wayland or not, so no need to worry about it for now. :)

Let me see if I can reproduce the issue on my side and we can go from there.

**Comment by josh-berry** [comment]
I unfortunately still haven't been able to reproduce this on my end, but once version 2.6 is released, let's re-test—I know there have been some changes in the drag-and-drop library used by Tab Stash, so hopefully this will resolve your issue.

**Comment by josh-berry** [comment]
Version 2.6 was just released, please let me know if you still see the issue on the latest version.

You can check which version of Tab Stash you have by scrolling to the very bottom of the stash list. If you're not on 2.6 yet, you can ask Firefox to update your add-ons by going to the add-ons page, opening the gear menu and choosing "Check for Updates".

**Comment by qqlovekerry** [comment]
Hi, I think I have updated to version 2.6 because I can see the message on the top of the tab view. But I think the issue still exists. It only happens when it is in the tab view. When I use the side view I do not see such an issue. Do you want me to make a video? However, I do not know how to upload a video on github.

**Comment by josh-berry** [comment]
Hm, sorry to hear that. A video would definitely help; I don't think GitHub allows attaching videos directly, but you could upload one to Firefox Send and put the link here. https://send.firefox.com/

I'll try to do some more reading on my end as well. The only thing I can think of is there's some weird interaction happening between the library I'm using for drag-and-drop and the handling for click events which is causing a click to be registered when it shouldn't be. But I still have no idea why it would only happen in the tab view.

**Comment by qqlovekerry** [comment]
Hi, I am sorry for the late response. I am sorry I do not the other ways to do so, so I have recorded a video with my cell phone. I hope it would be clear to see the problem. 
The link to find the video is https://send.firefox.com/download/e6c727cb93db9c40/#12rGOg9PB89pxW_c-WIoTA. I hope it helps.

**Comment by josh-berry** [comment]
No problem, thanks for taking the time to make the video! It was quite clear and I definitely see your problem, but I unfortunately have to confess I'm still not sure what's going on. I've tried installing an Ubuntu 18.04 VM just with default settings, and after installing updates it came up to Firefox 77.0.1. Still no luck in reproducing it here, but watching your video a few times gave me some more ideas for what to check.

First, some questions:
1. What Tab Stash options are you using? (check Tab Stash menu > Options)
2. What other Firefox extensions do you have installed?
3. What type of pointing device are you using? Mouse? Trackpad? Something else? (Do you have more than one pointing device, e.g. laptop with trackpad + an external mouse?)

And here are some experiments to try, in order from easy to hard:
1. In the tab view, try dragging a tab to and from "Unstashed Tabs". Does the tab open when it's dropped in "Unstashed Tabs", and does it close again when dragged back to the stash?
2. Try the same thing as (1), but in the sidebar view. Does it behave as expected?
3. Try moving tabs around without using the Tab Stash UI:
   1. Open the Firefox bookmarks manager (Firefox menu > Library > Bookmarks > Show All Bookmarks), and go to the "Other Bookmarks" > "Tab Stash" folder.
   2. Inside "Tab Stash" there will be a folder for each of your stashes (default stash names will look like "saved-2020-06-05T01:23:33Z"). Inside THOSE folders will be bookmarks for each of your tabs.
   3. Drag a bookmark from one of your stash folders to a different stash folder. Does the tab open when you do this?
4. Try disabling every other extension except Tab Stash, restart your browser, and repeat (1) and (2). Do you still see the problem?

If you still see the problem even with other extensions disabled, there's one more thing you can try, although I don't know if it will help: https://github.com/josh-berry/tab-stash/wiki/Restart-Firefox-Without-Saved-Session-Data

Let me know what you find, and thanks again for taking the time to help in troubleshooting. :)

**Comment by qqlovekerry** [comment]
Hi, thanks for the detailed reply! I followed your instructions and here are my tests.
For the questions:
1. I used `Don't automatically show the stash` and `Hide the tab and keep it loaded in the background`.
2. I have installed a bunch of addons, I am sorry I feel it is a bit difficult to list them all manually. Do you know some method to output them all quickly?
3. I am only using a mouse.
I also have some experiments as you suggested.
1. I recorded when I dragged the tab from and to the "Unstashed Tabs". And please download the video from the link <https://send.firefox.com/download/8df4d015b9a99a65/#YNlAyp4qROhhXJcF-bppeg>. I also found that when a group is dragged and dropped, it would open a tab in the browser with all the contents of the group searched in the default search engine.
2. In the sidebar view, I think the behavior is all good. All the problems happened only in the tab view.
3. If I move tabs in the Firefox books manager, no tabs would be opened in the browser.
4. I have lots of addons, some of them are already disabled, so disable or able them one by one manually seems a bit difficult for me. Can you point the problem already? Or could you please teach me some smart ways to do that?

**Comment by josh-berry** [comment]
Thanks for the video. It's interesting that no matter which way you're dragging something (into or out of Unstashed Tabs), it seems to be opening a new tab each time. I saw at one point you even had three of the same tab open. The fact that you're having the same problem with `start.me` suggests the problem might be with Firefox itself or another add-on. I'm wondering if there's some other extension or preference that is adding this new drag-and-drop behavior to all open tabs (but which doesn't take effect in the sidebar, which runs in a slightly different context in the browser).

To get a complete list of add-ons, including which ones are enabled/disabled, go to `about:support` and scroll down to "Extensions". You can copy/paste that whole table. Can you also send the "Important Modified Preferences" table from `about:support`? I want to see if there is anything unusual turned on there.

Unfortunately I don't know of a good way to disable all add-ons except Tab Stash. There is Firefox Safe Mode, but that will disable _all_ add-ons including Tab Stash, so I'm not sure it will help your testing. But if you save the "Extensions" list above, you can at least keep a note of which extensions were enabled/disabled before the experiment.

If you're uncomfortable sharing your preferences and extensions in a public bug report, you can email me directly at: `tab-stash AT condordes DOT net` (replacing AT and DOT accordingly). Just mention this issue (issue #51) in the subject so I know what your email is in reference to.

**Comment by josh-berry** [comment]
Hi, were you able to narrow this down any further?

**Comment by qqlovekerry** [comment]
> Hi, were you able to narrow this down any further?

Hi, I have sent the list of the installed extension on my Firefox to the email address tab-stash at condordes.net about one month ago. Did you receive the email?
But to disable all the addons and check which addon causes the problem seems to be so time-consuming, I am not able to do that yet.

**Comment by josh-berry** [comment]
Sorry, no I didn't receive it. :( I think there may have been some problem with spam filtering on my side, can you please try sending it again?

**Comment by qqlovekerry** [comment]
> Sorry, no I didn't receive it. :( I think there may have been some problem with spam filtering on my side, can you please try sending it again?

Hi, I have sent the email again. Could you please check whether the email has been received? If there is any problem, please tell me.

**Comment by josh-berry** [comment]
Got it, thank you! I'll take a look and get back to you.

**Comment by qqlovekerry** [comment]
> Got it, thank you! I'll take a look and get back to you.

Glad to hear that. I hope it helps. And thanks very much for your effort to solve this issue!
If there is anything I can do, please tell me. I will try my best to help.

**Comment by josh-berry** [comment]
My pleasure—thanks for your patience in trying to solve it! Hopefully we will catch it soon.

I read through your list of extensions and preferences, and unfortunately nothing stands out to me as potentially causing this... although I'm not familiar with a lot of the extensions you have installed, so I can't rule it out either. I did try installing a couple of extensions that seemed like they might cause conflicts (e.g. Gesturefy) but still wasn't able to reproduce the problem.

However, I did do some more searching around and ran across an open bug in the library I'm using to handle drag and drop (https://github.com/SortableJS/Sortable/issues/985), which sounds similar to your issue. It is possible that Firefox itself sees you are dragging a link, and for some reason decides to treat that drag operation as a request to open the link even though it is being handled by Tab Stash. I will put in a workaround for this bug for the next release. I'm not sure if this is actually your issue, so there are no guarantees, but hopefully this will help.

In the meantime, I do still think it would be useful to test without other extensions involved—if nothing else, this will help to rule out extensions and/or custom Firefox settings as the cause. I thought of another way you could do this that wouldn't be disruptive to your current Firefox, which is to install Firefox Developer Edition (which is a newer/beta version of Firefox that is completely isolated from your regular Firefox), and then install Tab Stash there. I think you can install Firefox Developer Edition directly through Ubuntu's software center, but if not, you can get it here: https://www.mozilla.org/en-US/firefox/developer/

If it still happens in the developer edition, that tells us this is an actual bug. But if not, it seems more likely to be a problem with some setting or extension in your browser (and you don't have to wait for the next release for a fix that wouldn't work for you anyway).

To summarize:
* I will put a workaround for the drag-and-drop bug mentioned above in the next release.
* If you can test Tab Stash in a separate/fresh install of Firefox Developer Edition, that will help us determine if that bug is actually the same as your problem.

**Comment by josh-berry** [comment]
[[note to self]]
* [ ] add `preventDefault()` to folder-list and folder drag events

**Comment by qqlovekerry** [comment]
Hi, I have downloaded the Firefox Developer Edition from the link you provided. I only installed the `Tab Stash` addon, and tested with drag and drop function. The problem does NOT happen in this mode. It behaves like expected. So, does this mean the problems are very likely caused by another addon on my firefox?

**Comment by josh-berry** [comment]
Oh interesting! A few possibilities seem most likely to me:
* It is the drag-and-drop bug in SortableJS I mentioned above, but there is some difference in behavior between Ubuntu and WIndows, and/or Firefox and DevEdition, that prevents the bug from happening
* There is a bug in Firefox itself which is fixed in Firefox Developer Edition/Firefox Beta
* There is some setting (which may not be synced) in your Firefox profile on Ubuntu that is causing the issue

Would you be open to trying Firefox Beta? The only difference between Beta and Developer Edition is Beta uses your existing profile/settings/add-ons. Otherwise they are the same version of Firefox (one version ahead of regular Firefox). If Beta has the problem but Developer Edition does not, that points us in the direction of your settings/add-ons. If both Beta and Developer Edition work OK, that points towards a Firefox bug (fixed in beta) or the SortableJS issue.

And if Firefox Beta works for you, hopefully that will resolve your problem in the meantime until I can get a new release done with the SortableJS workaround.

**Comment by qqlovekerry** [comment]
Hi, I think I have good news for us. Following your instructions and after intensive tests, I found the problem should be caused by one extension, it seems it only exists in Chinese language version Firefox, and I can not find it on the default add-ons search web site of Firefox (<https://addons.mozilla.org/>). It can be found here (<http://mozilla.com.cn/thread-343905-1-1.html>), but the web site is in Chinese and it is called `附加组件管理器`. If I disable this add-on, the "drag and drop" functions like expected, although I would lose some functions provided by that add-on.
The add-on caused the problem seems to be only provided for Chinese version, so it might hinder you or the community to find out the reason.
Another thing is that I memorize that on Windows, my Firefox also installed this add-on. However, I need to confirm that later when I use Windows on another PC of mine.

**Comment by josh-berry** [comment]
Awesome, glad to hear it! I took a quick look inside the extension, and it does seem to have some privileged access to Firefox to be able to control drag-and-drop events. Unfortunately there is not much I can do to control for this in Tab Stash (although I will still try to put the workaround and you can see if it helps).

The add-on seems to be distributed by Mozilla itself, so if you still want to use it, you might want to open a bug report with them.

Glad you were finally able to find the problem!

**Comment by qqlovekerry** [comment]
Hi, thanks for all your instructions! That helps a lot.
I am wondering does the same problem happen on your Firefox if you installed that add-on?

The add-on seems to be installed default on the Chinese version. But I am not able to find it on the official add-on website of Firefox. And I also cannot find the official website of that add-on. Could you please tell me how to find the distributor of the add-on and where to raise the issue?

**Comment by josh-berry** [comment]
From what I could see, the "add-on" is baked into the Chinese-language version of Firefox itself and is developed/distributed by Mozilla, so I think the right thing to do is to raise an issue with Mozilla.

**Comment by josh-berry** [comment]
So I checked just now and it seems the drag-and-drop library I'm using already does what I was thinking of doing for the workaround. I don't think there's anything more I can do here, so I'm going to go ahead and close this out.

Thanks again for all your help in troubleshooting. Hopefully Mozilla is able to address this in the Chinese version of Firefox, and sorry for the inconvenience!
